lib/rules/no-loop-func.js

```diff
@@ -186,13 +198,49 @@ module.exports = {
             }
 
             const references = sourceCode.getScope(node).through;
-            const unsafeRefs = references.filter(r => r.resolved && !isSafe(loopNode, r)).map(r => r.identifier.name);
+            const unsafeRefs = references.filter(r => r.resolved && !isSafe(loopNode, r));
+            const unsafeRefsName = unsafeRefs.map(r => r.identifier.name);
+
+            // Check if the function is not asynchronous or a generator function
+            if (!(node.async || node.generator)) {
+                const parentNode = node.parent;
+                const isFunctionExpression = node.type === "FunctionExpression";
+                const ArrowFunctionExpression = node.type === "ArrowFunctionExpression";
+
+                if (parentNode.type === "CallExpression" && (isFunctionExpression || ArrowFunctionExpression)) {
+
+                    // Check if the function is referenced elsewhere in the code
+                    const isFunctionReferenced = isFunctionExpression && node.id ? references.some(r => r.identifier.name === node.id.name) : false;
+
+                    // Check if there are unsafe references used within non-immediately-invoked nested functions
+                    const hasUnsafeRefsInNonIIFE = unsafeRefs.some(r => {
+                        let refScope = r.from.variableScope;
+
+                        if (node === refScope.block) {
+                            return !isIIFE(refScope.block);
+                        }
+
+                        while (node !== refScope.block) {
+                            if (!isIIFE(refScope.block)) {
+                                return true;
+                            }
+                            refScope = refScope.upper.variableScope;
+                        }
```

Comment:
This checks if the `node` itself is an IIFE only if the reference is directly in it. Since the `if (parentNode.type === "CallExpression" && (isFunctionExpression || ArrowFunctionExpression))` check is incomplete, there will be false negatives for references in inner functions when `node` is not IIFE but an argument of another IIFE:

```js
/* eslint no-loop-func: "error" */

var arr = [];

for (var i = 0; i < 5; i ++) {
    // false negative, reference to `i` is unsafe
    ((f) => {
        arr.push(f);
    })(() => {
        return (() => i)();
    });

}

console.log(arr.map(f => f())); // [ 5, 5, 5, 5, 5 ]

```

Reply:
Good catch, fixed!